crates/matrix-sdk/src/authentication/qrcode/messages.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use matrix_sdk_base::crypto::types::SecretsBundle;
+use openidconnect::{
+    core::CoreDeviceAuthorizationResponse, EndUserVerificationUrl, VerificationUriComplete,
+};
+use ruma::serde::StringEnum;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use url::Url;
+use vodozemac::Curve25519PublicKey;
+
+#[cfg(doc)]
+use crate::authentication::qrcode::QRCodeLoginError::SecureChannel;
+
+/// Messages that will be exchanged over the [`SecureChannel`] to log in a new
+/// device using a QR code.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(tag = "type")]
+pub enum QrAuthMessage {
+    /// Message declaring the available protocols for sign in. Sent by the
+    /// existing device.
+    #[serde(rename = "m.login.protocols")]
+    LoginProtocols {
+        /// The login protocols the existing device supports.
+        protocols: Vec<LoginProtocolType>,
+        /// The homeserver we're going to log in to.
+        homeserver: Url,
+    },
+
+    /// Message declaring which protocols from the previous `m.login.protocols`
+    /// message the new device has picked. Sent by the new device.
+    #[serde(rename = "m.login.protocol")]
+    LoginProtocol {
+        /// The device authorization grant the OIDC provider has given to the
+        /// new device, contains the URL the existing device should use
+        /// to confirm the log in.
+        device_authorization_grant: AuthorizationGrant,
+        /// The protocol the new device has picked.
+        protocol: LoginProtocolType,
+        #[serde(
+            deserialize_with = "deserialize_curve_key",
+            serialize_with = "serialize_curve_key"
+        )]
+        /// The device ID the new device will be using.
+        device_id: Curve25519PublicKey,
+    },
+
+    /// Message declaring that the protocol in the previous `m.login.protocol`
+    /// message was accepted. Sent by the extising device.
+    #[serde(rename = "m.login.protocol_accepted")]
+    LoginProtocolAccepted {},
+
+    /// Message that informs the existing device that it successfully obtained
+    /// an access token from the OIDC provider. Sent by the new device.
+    #[serde(rename = "m.login.success")]
+    LoginSuccess {},
```

Comment:
```diff
- LoginSuccess {},
+ LoginSuccess,
```

?

(and so on for the other variants of this enum)

Reply:
Duh: 191afaec22d059a94cc7ae54ed58f788a28cb198.